File: golang/Permutations.go
package main

import (
	"fmt"
)

func permute(s []rune, start int, result *[]string) {
	if start == len(s) {
		*result = append(*result, string(s))
		return
	}
	for i := start; i < len(s); i++ {
		s[start], s[i] = s[i], s[start]
		permute(s, start+1, result)
		s[start], s[i] = s[i], s[start] // backtrack
	}
}

func permutations(input string) []string {
	var result []string
	permute([]rune(input), 0, &result)
	return result
}

func main() {
	input := "abc"
	result := permutations(input)
	fmt.Println(result)
}

File: c-sharp/Permutations.cs
using System;
using System.Collections.Generic;

public class Permutations {
    private void Swap(ref char a, ref char b) {
        if (a == b) return;
        a ^= b;
        b ^= a;
        a ^= b;
    }

    private void GetPermutations(char[] list, int recursionDepth, ref List<string> result) {
        if (recursionDepth == list.Length) {
            result.Add(new string(list));
            return;
        }

        for (var i = recursionDepth; i < list.Length; i++) {
            Swap(ref list[recursionDepth], ref list[i]);
            GetPermutations(list, recursionDepth + 1, ref result);
            Swap(ref list[recursionDepth], ref list[i]);
        }
    }

    public List<string> Permute(string s) {
        var result = new List<string>();
        GetPermutations(s.ToCharArray(), 0, ref result);
        return result;
    }
    
    public static void Main() {
        var permutations = new Permutations();
        var result = permutations.Permute("abc");
        Console.WriteLine(string.Join(", ", result));
    }
}
